### ISSUE TITLE
Move from CFURLCreateStringByAddingPercentEscapes to stringByAddingPercentEncodingWithAllowedCharacters

### ISSUE THREAD

**Issue by josephbeuysmum** [mattdonnelly/Swifter#103]
Hello. The warnings are:

SwifterHTTPRequest.swift:170:31: 'init(request:delegate:)' was deprecated in iOS 9.0: Use NSURLSession (see NSURLSession.h)

and

String+Swifter.swift:57:22: 'CFURLCreateStringByAddingPercentEscapes' was deprecated in iOS 9.0: Use [NSString stringByAddingPercentEncodingWithAllowedCharacters:] instead, which always uses the recommended UTF-8 encoding, and which encodes for a specific URL component or subcomponent (since each URL component or subcomponent has different rules for what characters are valid).

I realise these are tiny problems, but I haven't done any programming since May 2015, and I feel that I might cause more harm than good by trying to fix them myself just yet, so just flagging them up for others until my mind is more Swift 2.0 attuned. JBM.

**Comment by alexdd55** [comment]
:+1:

**Comment by zntfdr** [comment]
Wops I missed this issue: otherwise I would have not opened a similar one ( https://github.com/mattdonnelly/Swifter/issues/119 ).

Let's split the two requests into two different issues:
in here we take care of the stringByAddingPercentEncodingWithAllowedCharacters shift while on mine ( https://github.com/mattdonnelly/Swifter/issues/119 ) we take care of the NSURLSession shift.

From what I can see `CFURLCreateStringByAddingPercentEscapes` is used only inside the `urlEncodedStringWithEncoding` function and everywhere in the framework this function is called with `NSUTF8StringEncoding` passed as the parameter value.

If my assumption is correct, I believe this issue can be solved by replacing the whole [urlEncodedStringWithEncoding](https://github.com/mattdonnelly/Swifter/blob/master/Swifter/String%2BSwifter.swift#L51) functino with a simpler: 
`````swift
func urlEncodedString() -> String {
  let allowedCharacters = NSCharacterSet(charactersInString: "[].")
  let str = self as NSString
  
  return  str.stringByAddingPercentEncodingWithAllowedCharacters(allowedCharacters)!
}
`````

Can somebody please give it a spin and see if it works? I'm out of town for work and I don't have a mac with me.

**Comment by lewis-smith** [comment]
Unless anyone will get to it beforehand, I'll take a look over the weekend

**Comment by zntfdr** [comment]
@lewis-smith awesome! Thanks :) :+1:

**Comment by meteochu** [comment]
I've been experimenting with URL Encoded Strings that work in (almost) all cases. It should work without any problem, as I've yet to encounter more since I last changed it... 
Simply add the following to a String extension and that'll do it!
```
var URLEncodedString: String {
        let allowedCharacterSet = NSCharacterSet.URLQueryAllowedCharacterSet().mutableCopy() as! NSMutableCharacterSet
        allowedCharacterSet.removeCharactersInString("\n:#[]/?@!$&'()*+,;=") // these characters need to be left alone :p
        return self.stringByAddingPercentEncodingWithAllowedCharacters(allowedCharacterSet)!
    }
```

**Comment by lewis-smith** [comment]
@zntfdr Your method encodes everything, including regular alpha characters, feel like that's gonna be tough for debugging (easy to miss when you're coding on the fly tho!)

@meteochu Your method looks good to me. But.. the original method doesn't encode [square braces] but yours does, regardless of whether you have the [] in the removeCharactersInString

@mattdonnelly I assume that there is a specific reason that [] were not encoded in the original code?

Aside: I've got the following test case:

```
let shouldBeTheSameAfterEncoding = [
            "a": "a",
            "a a": "a%20a",
            ".": ".",
            "[": "[",
            "]": "]",
            "'": "%27",
            "$": "%24",
            "&": "%26",
            "<": "%3C",
            ">": "%3E",
            "?": "%3F",
            ";": "%3B",
            "#": "%23",
            ":": "%3A",
            "=": "%3D",
            ",": "%2C",
            "\"": "%22",
            "+": "%2B",
            "%":  "%25",
            ":/?&=;+!@# ()',*": "%3A%2F%3F%26%3D%3B%2B%21%40%23%20%28%29%27%2C%2A",
            "http://a?a=[a,b]": "http%3A%2F%2Fa%3Fa%3D[a%2Cb]"
        ]

```

comments on it welcome.

**Comment by zntfdr** [comment]
I wrote my idea assuming that `stringByAddingPercentEncodingWithAllowedCharacters` was smart enough to leave regular chars not encoded, my bad. :sweat_smile: 

By looking at the repo history the whole function was added at https://github.com/mattdonnelly/Swifter/commit/debd07c2938a3b06d0ef724df77189493c3e3faa#diff-4c2f0a6ecfe49e40e378fd64b2e12a46 with comment "fixed percent escaping". 
It only changed once (https://github.com/mattdonnelly/Swifter/commit/1500f9cd03b1b4a7341678dbe644bbab9d0db8aa#diff-1ea5c1c04175aee133344e485fd0851f ), during its whole lifetime.

Also, the `encoding` parameter was present since the beginning, just wanted to point that out.

I also wonder why the `[ ]` shouldn't be encoded: from a simple test (using the new mac demo) I could see that only the dot "." has to be allowed (without it, the mac demo won't work). 

By removing the square brackets from `allowedCharacters` in my app (and also @meteochu 's , apparently) I couldn't notice any difference. (I launched a stream with only "[" and "]" as words to track and it worked flawlessly)

**Comment by meteochu** [comment]
The square brackets need to be ignored because of media uploads. The media upload parameter is `media[]`

**Comment by zntfdr** [comment]
@meteochu nice catch!

**Comment by lewis-smith** [comment]
Assuming we need square brackets, I have a fix. Will send a PR later in the week.

**Comment by zntfdr** [comment]
Solved in #135
